quickwit/quickwit-search/src/top_k_collector.rs

```diff
@@ -0,0 +1,860 @@
+    /// If the document is below the current threshold, it will be ignored.
+    #[inline]
+    pub fn push(&mut self, doc: D) {
+        if let Some(last_median) = self.threshold.clone() {
+            if doc < last_median {
+                return;
+            }
+        }
+        if self.buffer.len() == self.buffer.capacity() {
+            let median = self.truncate_top_n();
+            self.threshold = Some(median);
+        }
+
+        // This is faster since it avoids the buffer resizing to be inlined from vec.push()
+        // (this is in the hot path)
+        // TODO: Replace with `push_within_capacity` when it's stabilized
+        let uninit = self.buffer.spare_capacity_mut();
+        // This cannot panic, because we truncate_median will at least remove one element, since
+        // the min capacity is larger than 2.
+        uninit[0].write(Reverse(doc));
+        // This is safe because it would panic in the line above
+        unsafe {
+            self.buffer.set_len(self.buffer.len() + 1);
+        }
+    }
+
+    #[inline(never)]
+    fn truncate_top_n(&mut self) -> D {
+        // Use select_nth_unstable to find the top nth score
+        let (_, median_el, _) = self.buffer.select_nth_unstable(self.top_n);
+
+        let median_score = median_el.clone();
+        // Remove all elements below the top_n
+        self.buffer.truncate(self.top_n);
+
+        median_score.0
+    }
+
+    /// Returns the top n elements in sorted order.
+    pub fn into_sorted_vec(mut self) -> Vec<D> {
+        if self.buffer.len() > self.top_n {
+            self.truncate_top_n();
+        }
+        self.buffer.sort_unstable();
+        self.buffer.into_iter().map(|el| el.0).collect()
+    }
+
+    /// Returns the top n elements in stored order.
+    /// Useful if you do not need the elements in sorted order,
+    /// for example when merging the results of multiple segments.
+    #[allow(dead_code)]
+    pub fn into_vec(mut self) -> Vec<D> {
+        if self.buffer.len() > self.top_n {
+            self.truncate_top_n();
+        }
+        self.buffer.into_iter().map(|el| el.0).collect()
+    }
+}
+
+pub use tantivy::COLLECT_BLOCK_BUFFER_LEN;
+struct SpecSortingFieldExtractor<V1, V2> {
+    _phantom: std::marker::PhantomData<(V1, V2)>,
+    sort_values1: Box<[Option<u64>; COLLECT_BLOCK_BUFFER_LEN]>,
+    sort_values2: Box<[Option<u64>; COLLECT_BLOCK_BUFFER_LEN]>,
+
+    pub first: SortingFieldExtractorComponent,
+    pub second: Option<SortingFieldExtractorComponent>,
+}
+
+impl<
+        V1: Copy + PartialEq + PartialOrd + Ord + IntoOptionU64 + Debug,
+        V2: Copy + PartialEq + PartialOrd + Ord + IntoOptionU64 + Debug,
+    > SpecSortingFieldExtractor<V1, V2>
+{
+    fn new(
+        first: SortingFieldExtractorComponent,
+        second: Option<SortingFieldExtractorComponent>,
+    ) -> Self {
+        Self {
+            _phantom: PhantomData,
+            sort_values1: Box::new([None; COLLECT_BLOCK_BUFFER_LEN]),
```

Comment:
Nitpick.

Unfortunately this still allocates on the stack first. 
I tend to prefer `vec![None; COLLECT_BLOCK_BUFFER_LEN].into_boxed_slice()`?


In this very case this is only `9*64` bytes so it is probably ok.

Reply:
I thought this has been resolved, but it hasn't